src/native/tools/crypto_napi.cpp

```diff
@@ -0,0 +1,176 @@
+/* Copyright (C) 2016 NooBaa */
+#include <string.h>
+#include <vector>
+#include "../third_party/isa-l_crypto/include/md5_mb.h"
+#include "../util/common.h"
+#include "../util/endian.h"
+#include "../util/napi.h"
+
+namespace noobaa
+{
+
+template <typename T>
+static Napi::Value
+api(const Napi::CallbackInfo& info)
+{
+    auto w = new T(info);
+    Napi::Promise promise = w->_deferred.Promise();
+    w->Queue();
+    return promise;
+}
+
+/**
+ * CryptoWorker is a general async worker for our crypto operations
+ */
+struct CryptoWorker : public Napi::AsyncWorker
+{
+    Napi::Promise::Deferred _deferred;
+
+    CryptoWorker(const Napi::CallbackInfo& info)
+        : AsyncWorker(info.Env())
+        , _deferred(Napi::Promise::Deferred::New(info.Env()))
+    {
+    }
+    virtual void OnOK() override
+    {
+        // LOG("CryptoWorker::OnOK: undefined");
+        _deferred.Resolve(Env().Undefined());
+    }
+    virtual void OnError(Napi::Error const& error) override
+    {
+        LOG("CryptoWorker::OnError: " << DVAL(error.Message()));
+        auto obj = error.Value();
+        _deferred.Reject(obj);
+    }
+};
+
+/**
+ * CryptoWrapWorker is meant to simplify adding async CryptoWorker instance methods to ObjectWrap types
+ * like MD5Wrap, while keeping the object referenced during that action.
+ */
+template <typename Wrapper>
+struct CryptoWrapWorker : public CryptoWorker
+{
+    Wrapper* _wrap;
+    CryptoWrapWorker(const Napi::CallbackInfo& info)
+        : CryptoWorker(info)
+    {
+        _wrap = Wrapper::Unwrap(info.This().As<Napi::Object>());
+        _wrap->Ref();
+    }
+    ~CryptoWrapWorker()
+    {
+        _wrap->Unref();
+    }
+};
+
+struct MD5Wrap : public Napi::ObjectWrap<MD5Wrap>
+{
+    size_t _NWORDS = MD5_DIGEST_NWORDS;
+    bool _WORDS_BE = false;
+    void (*_INIT)(MD5_HASH_CTX_MGR*) = md5_ctx_mgr_init;
+    MD5_HASH_CTX *(*_SUBMIT)(MD5_HASH_CTX_MGR*, MD5_HASH_CTX*, const void*, uint32_t, HASH_CTX_FLAG) = md5_ctx_mgr_submit;
+    MD5_HASH_CTX *(*_FLUSH)(MD5_HASH_CTX_MGR*) = md5_ctx_mgr_flush;
+    DECLARE_ALIGNED(MD5_HASH_CTX_MGR _mgr, 16);
+    DECLARE_ALIGNED(MD5_HASH_CTX _ctx, 16);
```

Comment:
We can have `typedef Hasher` same as we have it here:
https://github.com/noobaa/noobaa-core/blob/8aa2d394fb716f060d950594ae720f3674d41318/src/native/tools/ssl_napi.cpp#L130-L139

Reply:
Right, I don't really understand why you would copy the Hasher template code and then remove the templating and then have just md5 instead of any type?

Reply:
Also, I think that we should use OpenSSL by default and only fallback to ISA-L for fips mode.

Reply:
We can do that in the subsequent PR after the MVP scope.
For the MVP we need MD5 solely.
Did not add SHA1 in order to minimize the footprint and deliver today, also not to mess with FIPS.